components/cloud/gcp/src/gcs.rs

```diff
@@ -345,6 +364,86 @@ impl BlobStorage for GcsStorage {
     }
 }
 
+struct GcsWalker<'cli, 'arg> {
+    cli: &'cli GcsStorage,
+    page_token: Option<String>,
+    prefix: &'arg str,
+    finished: bool,
+}
+
+impl<'cli, 'arg> GcsWalker<'cli, 'arg> {
+    async fn one_page(&mut self) -> io::Result<Option<Vec<BlobObject>>> {
+        if self.finished {
+            return Ok(None);
+        }
+
+        let mut opt = ListOptional::default();
+        let bucket =
+            BucketName::try_from(self.cli.config.bucket.bucket.to_string()).or_invalid_input(
+                format_args!("invalid bucket {}", self.cli.config.bucket.bucket),
+            )?;
+        let prefix = self.cli.maybe_prefix_key(self.prefix);
+        opt.prefix = Some(&prefix);
+        opt.max_results = Some(2048);
+        opt.page_token = self.page_token.as_deref();
+        let req = Object::list(&bucket, Some(opt)).or_io_error(format_args!(
+            "failed to list with prefix {} page_token {:?}",
+            self.prefix, self.page_token
+        ))?;
+        let res = self
+            .cli
+            .make_request(req.map(|_e| Body::empty()), tame_gcs::Scopes::ReadOnly)
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+        let resp = utils::read_from_http_body::<ListResponse>(res).await?;
+        debug!("requesting paging GCP"; "prefix" => self.prefix, "page_token" => self.page_token.as_deref(), 
+            "response_size" => resp.objects.len(), "new_page_token" => resp.page_token.as_deref());
+        // GCP returns an empty page token when returning the last page...
+        // We need to break there or we will enter an infinity loop...
+        if resp.page_token.is_none() || resp.objects.is_empty() {
```

Comment:
we probably should not check on the `objects.isEmpty()`. There is a case that objects could be empty but token is not. It could spend a great amount of time trying to search for the required items but none is found, in that case it might return early with a token to prevent request timeout and you can use the token to continue the search until boundary is reached

The reason why it can time out is it might contain a lot of tombstone or outdated version objects, and all the time is spent on skipping them